Add some docs about deploy tempo Gateway and use it with Grafana (#3135)
grafana/tempo: docs/sources/tempo/operations/grafana_datasource.md

---
title: Grafana Tempo Datasource

> description: A collection of documents that detail how to deploy tempo and use it with Grafana

Reviewer: Is this a correct summary of the the goal of these pages? 

```suggestion
description: Use the Tempo Operator to deploy Tempo and use it as a data source with Grafana
```

aliases:
 - /docs/tempo/latest/gateway
 - /docs/tempo/operations/gateway
weight: 15
---

# Grafana Tempo Datasource

> There are one main thing to take into account when you deploy the `TempoStack` and want to use as Grafana datasource.

Reviewer: ```suggestion
When you deploy the `TempoStack` and want to use as Tempo as a Grafana data source, you'll need to provide some additional information.
```

> # Grafana Tempo Datasource
> 
> There are one main thing to take into account when you deploy the `TempoStack` and want to use as Grafana datasource.
> 
> If the tempo is deployed the Gateway in front, an authentication information needs to be provided to Grafana, and the URL of the tenant it wants to access.
> 
> If the gateway is not used, the only requirement is to Grafana can access to the `frontend-query` endpoints.

Reviewer: I think the first few sentences in this page are really discussing additional things a user needs to have or do. 

The intro uses `frontend-query` and the Without the gateway section uses `query-frontend`. Which one is correct? 

```suggestion
# Grafana Tempo Datasource

Using the instructions on this page, you can configure the `TempoStack` to send data to Grafana and configure the Tempo data source.  

You can choose to either use the gateway or not use the gateway: 

* If the `TempoStack` is deployed using the Gateway in front, you'll need to provide authentication information  to Grafana, along with the URL of the tenant it wants to access.

* If the gateway is not used, then you need to make sure Grafana can access to the `frontend-query` endpoints.
```

Author: query-frontend should be the right one

## Use with gateway

> The gateway is an optional component deployed as part of Tempo Operator. It provides secure access to Tempo's distributor (i.e. for pushing spans) and query-frontend (i.e. for querying traces) via consulting an OAuth/OIDC endpoint for the request subject.
> 
> The OIDC configuration expects `clientID`, `clientSecret`  which should be provided via a Kubernetes secret that the TempoStack admin provides upfront.
> 
> The gateway exposes all tempo query endpoints, thus we can use the endpoint as a Tempo Grafana Datasource.
> 
> If  Grafana is configured with some OAuth provider, e.g. generic oauth, TempoStack with the gateway should be deployed  using the same `clientID` and `clientSecret`:

Reviewer: ```suggestion
The gateway, an optional component deployed as part of Tempo Operator, provides secure access to Tempo's distributor (for example, for pushing spans) and query-frontend (for example, for querying traces) via consulting an OAuth/OIDC endpoint for the request subject.

The OIDC configuration expects `clientID`, `clientSecret` which should be provided via a Kubernetes secret that the `TempoStack` admin provides upfront.

The gateway exposes all Tempo query endpoints, so you can use the endpoint as a Tempo Grafana data source.

If Grafana is configured with some OAuth provider, such as generic OAuth, the `TempoStack` with the gateway should be deployed using the same `clientID` and `clientSecret`:
```

```yaml
apiVersion: v1
kind: Secret
metadata:
 name: oidc-test
stringData:
 clientID: <clientID used for grafana authentication>
 clientSecret: <clientSecret used for grafana authentication>
type: Opaque
```

> Then deploy TempoStack with gateway enabled

Reviewer: ```suggestion
Then deploy TempoStack with gateway enabled:
```

```yaml
spec:
 template:
  gateway:
   enabled: true
 tenants:
  mode: static
  authentication:
    - tenantName: test-oidc
      tenantId: test-oidc
      oidc:
      issuerURL: http://dex:30556/dex
      redirectURL: http://tempo-foo-gateway:8080/oidc/test-oidc/callback
      usernameClaim: email
      secret:
       name: oidc-test
```

> In the url you should set it to `http://<HOST>:<PORT>/api/traces/v1/{tenant}/tempo/`

Reviewer: When you say URL, do you mean the issuerURL or the redirectURL or something else?

Author: I mean the URL on the datasource, on the grafana datasource field.

Author: I rephrased this to be more clear. Thanks!

> In order to use it as a data source you need to set the Authentication Method to **Forward Oauth Identify**, and make sure  you use the same `clientID` and `clientSecret` for gateway and for the oauth configuration. This will forward the access_token to the gateway ,so it can authenticate the client.

Reviewer: ```suggestion
To use it as a data source, set the Authentication Method to **Forward Oauth Identify** using the same `clientID` and `clientSecret` for gateway and for the OAuth configuration. This will forward the `access_token` to the gateway so it can authenticate the client.
```

<p align="center"><img src="../grafana_datasource_tempo.png" alt="Tempo Datasource configured for the gateway forwarding oauth access token"></p>

> If you prefer to set the Bearer token directly, and not use the  **Forward Oauth Identify** you can add it to the "Authorization" Header 

Reviewer: ```suggestion
If you prefer to set the Bearer token directly and not use the  **Forward Oauth Identify**, you can add it to the "Authorization" Header 
```

<p align="center"><img src="../grafana_datasource_tempo_headers.png" alt="Tempo Datasource configured for the gateway using Bearer token"></p>

## Without the gateway

> If you are not using the gateway make sure your grafana can access to the query-frontend endpoints, you can do this 
> creating an ingress or a route in OpenShift.
> 
> Once you have the endpoint you can set it as `URL` when you create the tempo datasource.

Reviewer: ```suggestion
If you are not using the gateway, make sure your Grafana can access to the query-frontend endpoints, you can do this by creating an ingress or a route in OpenShift.

Once you have the endpoint, you can set it as `URL` when you create the Tempo data source.
```

---
title: Grafana Tempo Datasource
description: Use the Tempo Operator to deploy Tempo and use it as a data source with Grafana
aliases:
 - /docs/tempo/latest/gateway
 - /docs/tempo/operations/gateway
weight: 15
---

> # Grafana Tempo Datasource

Reviewer: Data source is usually used as two words. 

```suggestion
# Grafana Tempo data source
```

Using the instructions on this page, you can configure the `TempoStack` to send data to Grafana and configure the Tempo data source.  

You can choose to either use Tempo Operator's gateway or not: 

> * If the `TempoStack` is deployed using the gateway, you'll need to provide authentication information  to Grafana, along with the URL of the tenant it wants to access.

Reviewer: In this phrase ("along with the URL of the tenant it wants to access"), "it" refers to tenant (the noun that precedes it). Do you mean tenant or the gateway? If the latter is correct, we should change "it" to gateway.

Author: none of them, I would like to express that you need to provide the URL of the tenant  you want to see the traces.

Author: Rephrased to make it more clear, let me know what do you think.

* If the gateway is not used, then you need to make sure Grafana can access to the `query-frontend` endpoints.

## Use with gateway

The gateway, an optional component deployed as part of Tempo Operator, provides secure access to Tempo's distributor (for example, for pushing spans) and query-frontend (for example, for querying traces) via consulting an OAuth/OIDC endpoint for the request subject.

> The OIDC configuration expects `clientID`, `clientSecret` which should be provided via a Kubernetes secret that the `TempoStack` admin provides upfront.

Reviewer: Is this correct that both clientID and clientSecret are provided via K8s secret or is it only clientSecret?
```suggestion
The OIDC configuration expects `clientID` and `clientSecret`. They should be provided via a Kubernetes secret that the `TempoStack` admin provides upfront.
```

Author: both are provided via Secret

The gateway exposes all Tempo query endpoints, so you can use the endpoint as a Tempo Grafana data source.

If Grafana is configured with some OAuth provider, such as generic OAuth, the `TempoStack` with the gateway should be deployed using the same `clientID` and `clientSecret`:

```yaml
apiVersion: v1
kind: Secret
metadata:
 name: oidc-test
stringData:
 clientID: <clientID used for grafana authentication>
 clientSecret: <clientSecret used for grafana authentication>
type: Opaque
```

Then deploy `TempoStack` with gateway enabled:

```yaml
spec:
 template:
  gateway:
   enabled: true
 tenants:
  mode: static
  authentication:
    - tenantName: test-oidc
      tenantId: test-oidc
      oidc:
      issuerURL: http://dex:30556/dex
      redirectURL: http://tempo-foo-gateway:8080/oidc/test-oidc/callback
      usernameClaim: email
      secret:
       name: oidc-test
```

> The datasource URL parameter should set to `http://<HOST>:<PORT>/api/traces/v1/{tenant}/tempo/`

Reviewer: ```suggestion
Set the data source URL parameter to: `http://<HOST>:<PORT>/api/traces/v1/{tenant}/tempo/`.
```

To use it as a data source, set the Authentication Method to **Forward Oauth Identify** using the same `clientID` and `clientSecret` for gateway and for the OAuth configuration. This will forward the `access_token` to the gateway so it can authenticate the client.

> <p align="center"><img src="../grafana_datasource_tempo.png" alt="Tempo Datasource configured for the gateway forwarding oauth access token"></p>

Reviewer: ```suggestion
<p align="center"><img src="../grafana_datasource_tempo.png" alt="Tempo data source configured for the gateway forwarding OAuth access token"></p>
```

---

> title: Grafana Tempo Datasource
> description: Use the Tempo Operator to deploy Tempo and use it as a Datasource with Grafana

Reviewer: ```suggestion
title: Grafana Tempo data source
description: Use the Tempo Operator to deploy Tempo and use it as a data source with Grafana
```

aliases:
 - /docs/tempo/latest/gateway
 - /docs/tempo/operations/gateway
weight: 15
---

> # Grafana Tempo Datasource

Reviewer: The Grafana UI does not capitalize data source as as proper noun, so I've changed all occurances of data source to lower case. 
![image](https://github.com/grafana/tempo/assets/104772500/03ba5610-e60c-4525-9546-9a5bd3e19544)

```suggestion
# Grafana Tempo data source
```

> Using the instructions on this page, you can configure the `TempoStack` to send data to Grafana and configure the Tempo Datasource.  

Reviewer: ```suggestion
Using the instructions on this page, you can configure the `TempoStack` to send data to Grafana and configure the Tempo data source.  
```

> Using the instructions on this page, you can configure the `TempoStack` to send data to Grafana and configure the Tempo data source.  

Reviewer: ```suggestion
You can configure the `TempoStack` to send data to Grafana and configure the Tempo data source.  
```

You can choose to either use Tempo Operator's gateway or not: 

* If the `TempoStack` is deployed using the gateway, you'll need to provide authentication information to Grafana, along with the URL of the tenant from which you expect to see the traces.

* If the gateway is not used, then you need to make sure Grafana can access the `query-frontend` endpoints.

> For more information, refer to the [Tempo Datasource for Grafana](/docs/grafana/latest/datasources/tempo/),

Reviewer: ```suggestion
For more information, refer to the [Tempo data source for Grafana](/docs/grafana/latest/datasources/tempo/),
```

## Use with gateway

The gateway, an optional component deployed as part of Tempo Operator, provides secure access to Tempo's distributor (for example, for pushing spans) and query-frontend (for example, for querying traces) via consulting an OAuth/OIDC endpoint for the request subject.

The OIDC configuration expects `clientID` and `clientSecret`. They should be provided via a Kubernetes secret that the `TempoStack` admin provides upfront.

> The gateway exposes all Tempo query endpoints, so you can use the endpoint as a Tempo Grafana Datasource.

Reviewer: ```suggestion
The gateway exposes all Tempo query endpoints, so you can use the endpoint as a Tempo data source for Grafana.
```

If Grafana is configured with some OAuth provider, such as generic OAuth, the `TempoStack` with the gateway should be deployed using the same `clientID` and `clientSecret`:

```yaml
apiVersion: v1
kind: Secret
metadata:
 name: oidc-test
stringData:
 clientID: <clientID used for grafana authentication>
 clientSecret: <clientSecret used for grafana authentication>
type: Opaque
```

Then deploy `TempoStack` with gateway enabled:

```yaml
spec:
 template:
  gateway:
   enabled: true
 tenants:
  mode: static
  authentication:
    - tenantName: test-oidc
      tenantId: test-oidc
      oidc:
      issuerURL: http://dex:30556/dex
      redirectURL: http://tempo-foo-gateway:8080/oidc/test-oidc/callback
      usernameClaim: email
      secret:
       name: oidc-test
```

> Set the Datasource URL parameter to `http://<HOST>:<PORT>/api/traces/v1/{tenant}/tempo/`, where `{tenant}` is the name of the tenant.

Reviewer: ```suggestion
Set the data source URL parameter to `http://<HOST>:<PORT>/api/traces/v1/{tenant}/tempo/`, where `{tenant}` is the name of the tenant.
```

> To use it as a Datasource, set the Authentication Method to **Forward Oauth Identify** using the same `clientID` and `clientSecret` for gateway and for the OAuth configuration. This will forward the `access_token` to the gateway so it can authenticate the client.

Reviewer: In the Grafana UI, data source is two words and lowercase most of the time. We should use the same reference in our documentation. 

```suggestion
To use it as a data source, set the Authentication Method to **Forward Oauth Identify** using the same `clientID` and `clientSecret` for gateway and for the OAuth configuration. This will forward the `access_token` to the gateway so it can authenticate the client.
```

> <p align="center"><img src="../grafana_datasource_tempo.png" alt="Tempo Datasource configured for the gateway forwarding OAuth access token"></p>

Reviewer: ```suggestion
<p align="center"><img src="../grafana_datasource_tempo.png" alt="Tempo data source configured for the gateway forwarding OAuth access token"></p>
```

> <p align="center"><img src="../grafana_datasource_tempo.png" alt="Tempo Datasource configured for the gateway forwarding OAuth access token"></p>
> 
> If you prefer to set the Bearer token directly and not use the  **Forward Oauth Identify**, you can add it to the "Authorization" Header.
> 
> <p align="center"><img src="../grafana_datasource_tempo_headers.png" alt="Tempo Datasource configured for the gateway using Bearer token"></p>

Reviewer: Same issue as above. 

```suggestion
<p align="center"><img src="../grafana_datasource_tempo.png" alt="Tempo data source configured for the gateway forwarding OAuth access token"></p>

If you prefer to set the Bearer token directly and not use the  **Forward Oauth Identify**, you can add it to the "Authorization" Header.

<p align="center"><img src="../grafana_datasource_tempo_headers.png" alt="Tempo data source configured for the gateway using Bearer token"></p>
```

## Without the gateway

If you are not using the gateway, make sure your Grafana can access to the query-frontend endpoints, you can do this by creating an ingress or a route in OpenShift.

> Once you have the endpoint, you can set it as `URL` when you create the Tempo Datasource.

Reviewer: ```suggestion
Once you have the endpoint, you can set it as `URL` when you create the Tempo data source.
```
